### PR TITLE
Add Range[Inclusive]::is_empty

### PR DESCRIPTION
**Pull request by scottmcm**
During https://github.com/rust-lang/rfcs/pull/1980, it was discussed that figuring out whether a range is empty was subtle, and thus there should be a clear and obvious way to do it.  It can't just be ExactSizeIterator::is_empty (also unstable) because not all ranges are ExactSize -- such as `Range<i64>` and `RangeInclusive<usize>`.

Things to ponder:
- Unless this is stabilized first, this makes stabilizing ExactSizeIterator::is_empty more icky, since this hides that.
- This is only on `Range` and `RangeInclusive`, as those are the only ones where it's interesting.  But one could argue that it should be on more for consistency, or on RangeArgument instead.
- The bound on this is PartialOrd, since that works ok (see tests for float examples) and is consistent with `contains`.  But ranges like `NAN..=NAN`_are_ kinda weird.
- [x] ~~There's not a real issue number on this yet~~

### CODE REVIEW

File: src/libcore/iter/traits.rs [hunk 1]
```diff
@@ -706,7 +706,7 @@ pub trait ExactSizeIterator: Iterator {
     /// ```
     /// #![feature(exact_size_is_empty)]
     ///
-    /// let mut one_element = 0..1;
+    /// let mut one_element = std::iter::once(0);
```
Comment (shepmaster) [inline src/libcore/iter/traits.rs]: Why is this related to this PR?

File: src/libcore/tests/iter.rs [hunk 2]
```diff
@@ -1427,9 +1427,9 @@ fn test_range_inclusive_nth() {
     assert_eq!(r, 13..=20);
     assert_eq!(r.nth(2), Some(15));
     assert_eq!(r, 16..=20);
-    assert_eq!(r.is_empty(), false);
+    assert_eq!(ExactSizeIterator::is_empty(&r), false);
     assert_eq!(r.nth(10), None);
-    assert_eq!(r.is_empty(), true);
+    assert_eq!(ExactSizeIterator::is_empty(&r), true);
```
Comment (shepmaster) [inline src/libcore/tests/iter.rs]: Is this going to be a concern for end users?

**Comment by shepmaster** [comment]
The code seems reasonable to me, but since @kennytm has been thinking about ranges lately...

r? @kennytm

File: src/libcore/tests/ops.rs [hunk 3]
```diff
@@ -68,3 +68,27 @@ fn test_range_inclusive() {
...
+
+    assert!(!(0.0 ..= 10.0).is_empty());
+    assert!(!(-0.0 ..= 0.0).is_empty());
+    assert!( (10.0 ..= 0.0).is_empty());
+
+    assert!(!(NEG_INFINITY ..= INFINITY).is_empty());
+    assert!( (EPSILON ..= NAN).is_empty());
+    assert!( (NAN ..= EPSILON).is_empty());
+    assert!( (NAN ..= NAN).is_empty());
+}
```
Comment (shepmaster) [inline src/libcore/tests/ops.rs]: WDYT about a test that calls `next` on a `Range{Inclusive}` and makes sure it transitions from not  empty to empty?

File: src/libcore/iter/traits.rs (hunk 1 above)
Comment (scottmcm) [reply src/libcore/iter/traits.rs]: Because using `0..1` means the code wasn't actually using `ExactSizeIterator`; the inherent method was getting picked up instead.  For normal code that's not a problem--they do exactly the same thing--but I figured the doctest should actually use the method it's documenting.

File: src/libcore/tests/iter.rs (hunk 2 above)
Comment (shepmaster) [reply src/libcore/tests/iter.rs]: > ExactSizeIterator::is_empty (also unstable)

ah, nvm

Comment (scottmcm) [reply src/libcore/tests/iter.rs]: `Range::is_empty` is strictly more general, so it's not a problem in a world where everything is stable.  It may be a slight hiccup in nightly, but I would expect it to be solved in the wild by adding the new feature gate, rather than the annoying UFCS.  Here I just changed it to the other method because I didn't want to change what the test was testing.  (It _would_ be a problem if we wanted to stabilize `ExactSizeIterator::is_empty` first, as mentioned in the OP, but I feel like this should be less controversial than the new trait method, especially since there are a bunch of iterators that aren't `ExactSize` that could provide an efficient `is_empty`.)

File: src/libcore/tests/ops.rs (hunk 3 above)
Comment (scottmcm) [reply src/libcore/tests/ops.rs]: Good thought; I'll change all of these to look for is_empty instead of `1..=0`: https://github.com/rust-lang/rust/blob/master/src/libcore/tests/iter.rs#L1433

File: src/libcore/tests/iter.rs [hunk 4]
```diff
@@ -1428,9 +1470,10 @@ fn test_range_inclusive_nth() {
     assert_eq!(r.nth(2), Some(15));
     assert_eq!(r, 16..=20);
     assert_eq!(r.is_empty(), false);
+    assert_eq!(ExactSizeIterator::is_empty(&r), false);
     assert_eq!(r.nth(10), None);
     assert_eq!(r.is_empty(), true);
-    assert_eq!(r, 1..=0);  // We may not want to document/promise this detail
+    assert_eq!(ExactSizeIterator::is_empty(&r), true);
```
Comment (scottmcm) [inline src/libcore/tests/iter.rs]: Changed RangeInclusive exhaustion tests to check `is_empty`, updated the docs to unspecify the post-iteration value of a RangeInclusive, and added some more Range tests using its `is_empty`.

**Comment by kennytm** [comment]
Since you're defining `is_empty()` to floating point types as well, perhaps spell out in the documentation that `a..b` is not empty when `a < b`.

LGTM other than this documentation bit, but I'd also like someone from @rust-lang/libs to sign-off.

**Comment by alexcrichton** [comment]
Looks good to me! Especially as unstable I think this can land when it's good to go

**Comment by scottmcm** [comment]
Added the `start < end` and `start <= end` to the doc comments next to the description of what the range contains, and added a doctest for each showing the behaviour with `NAN`.
